### PR TITLE
MID generator feature

### PR DESCRIPTION
**Pull request by algesten**

### CODE REVIEW

File: src/api/setting_engine/mod.rs [hunk 1]
```diff
@@ -74,6 +74,7 @@ pub struct SettingEngine {
     pub(crate) disable_media_engine_copy: bool,
     pub(crate) srtp_protection_profiles: Vec<SrtpProtectionProfile>,
     pub(crate) receive_mtu: usize,
+    pub(crate) mid_generator: Option<Arc<Box<dyn Fn() -> String + Send + Sync>>>,
```
Comment (k0nserv) [inline src/api/setting_engine/mod.rs:77]: I don't think we need `Box` here? `Arc<dyn Fn() -> Strign + Send + Sync>` ought to be enough.

File: src/api/setting_engine/mod.rs [hunk 2]
```diff
@@ -294,4 +295,11 @@ impl SettingEngine {
     pub fn set_receive_mtu(&mut self, receive_mtu: usize) {
         self.receive_mtu = receive_mtu;
     }
+
+    /// Sets a callback used to generate mid for transceivers create by this side of the RTCPeerconnection.
+    /// By having separate "naming schemes" for mids generated by either side of a connection, it's
+    /// possible to reduce complexity when handling SDP offers/answers clashing.
+    pub fn set_mid_generator(&mut self, f: impl Fn() -> String + Send + Sync + 'static) {
```
Comment (k0nserv) [inline src/api/setting_engine/mod.rs:302]: I think the callback should take an isize argument which is the value of greater_mid value at the time of calling it

Comment (lolgesten) [reply src/api/setting_engine/mod.rs:302]: sure

File: src/api/setting_engine/mod.rs (hunk 1 above)
Comment (k0nserv) [reply src/api/setting_engine/mod.rs:77]: I think let's leave it as sync until someone complains?

Comment (algesten) [reply src/api/setting_engine/mod.rs:77]: alright

Comment (algesten) [reply src/api/setting_engine/mod.rs]: done

File: src/peer_connection/mod.rs [hunk 3]
```diff
@@ -765,8 +765,28 @@ impl RTCPeerConnection {
...
+
+                        // If it's possible to parse the returned mid as numeric, we will update the greater_mid field.
+                        let numeric_mid = match mid.parse::<isize>() {
+                            Ok(n) => n,
+                            Err(_) => continue,
+                        };
+
+                        t.set_mid(mid).await?;
+
+                        if numeric_mid > self.internal.greater_mid.load(Ordering::SeqCst) {
```
Comment (k0nserv) [inline src/peer_connection/mod.rs:781]: This isn't done in the other case why do we need to do it here 🤔

Comment (lolgesten) [reply src/peer_connection/mod.rs:781]: See above for incoming mids

**Review by k0nserv — APPROVED** [review]

File: src/api/setting_engine/mod.rs [hunk 4]
```diff
@@ -294,4 +295,18 @@ impl SettingEngine {
     pub fn set_receive_mtu(&mut self, receive_mtu: usize) {
         self.receive_mtu = receive_mtu;
     }
+
+    /// Sets a callback used to generate mid for transceivers create by this side of the RTCPeerconnection.
```
Comment (k0nserv) [inline src/api/setting_engine/mod.rs:299]: `create` –> `created`

**Pull request merged**